enclave/core/sgx/once.c

```diff
@@ -5,34 +5,73 @@
 #include <openenclave/internal/thread.h>
 #include <openenclave/internal/utils.h>
 
+#define FUNC_NOT_INVOKED 0
+#define FUNC_BEING_INVOKED 1
+#define FUNC_INVOKED 2
+
 oe_result_t oe_once(oe_once_t* once, void (*func)(void))
 {
     if (!once)
         return OE_INVALID_PARAMETER;
 
     /* Double checked locking (DCLP). */
-    oe_once_t o = *once;
-
     /* DCLP Acquire barrier. */
     OE_ATOMIC_MEMORY_BARRIER_ACQUIRE();
```

Comment:
What does the memory barrier do in the new code? The original code had a `oe_once_t o = *once` statement before so the memory barrier would prevent the statement reordering.

Reply:
It makes sure that all the writes that happen in other threads before write to the *once variable happen (i.e observed by this thread), before the *once variable is read by this thread.

A good reference is C++11 memory ordering which describes the semantics more formally https://en.cppreference.com/w/cpp/atomic/memory_order:

> memory_order_acquire A load operation with this memory order performs the acquire operation on the affected memory location: no reads or writes in the current thread can be reordered before this load. All writes in other threads that release the same atomic variable are visible in the current thread (see Release-Acquire ordering below)

> memory_order_release : no reads or writes in the current thread can be reordered after this store. All writes in the current thread are visible in other threads that acquire the same atomic variable (see Release-Acquire ordering below) and writes that carry a dependency into the atomic variable become visible in other threads that consume the same atomic (see Release-Consume ordering below).